test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
+            assert.strictEqual(numTimesReleased, 1);
 
-      beforeEach(function() {
-        fakeSession = {
-          type: 'readwrite',
-        };
+            requestStream.abort();
+            assert.strictEqual(numTimesReleased, 1);
 
-        sessionPool.borrowed_ = [fakeSession];
-      });
+            done();
+          });
+        });
 
-      it('should create a new transaction', function(done) {
-        sessionPool.createTransaction_ = function(session) {
-          assert.strictEqual(session, fakeSession);
-          return Promise.resolve();
-        };
+        it('should cancel the request stream', function(done) {
+          REQUEST_STREAM.cancel = done;
 
-        sessionPool.release_ = function(session) {
-          assert.strictEqual(session, fakeSession);
-          done();
-        };
+          const requestStream = sessionPool.requestStream(CONFIG);
 
-        sessionPool.release(fakeSession);
+          requestStream.emit('reading');
+
+          setImmediate(function() {
+            requestStream.abort();
+          });
+        });
       });
+    });
 
-      it('should set to readonly if it fails to create a txn', function(done) {
-        sessionPool.createTransaction_ = function() {
-          return Promise.reject();
+    describe('createTransaction_', function() {
+      let fakeSession;
+      let fakeTxn;
+
+      beforeEach(function() {
+        fakeTxn = {
+          end: function(callback) {
+            (common.util.noop || callback)();
+          },
+          begin: function() {
+            return Promise.resolve();
+          },
         };
 
-        sessionPool.release_ = function(session) {
-          assert.strictEqual(session, fakeSession);
-          assert.strictEqual(fakeSession.type, 'readonly');
-          done();
+        fakeSession = {
+          transaction: function() {
+            return fakeTxn;
+          },
         };
 
-        sessionPool.release(fakeSession);
+        sessionPool.release = common.util.noop;
       });
-    });
-  });
-
-  describe('request', function() {
-    var CONFIG;
-    var SESSION = {
-      formattedName_: 'formatted-name',
-    };
-
-    beforeEach(function() {
-      CONFIG = {
-        reqOpts: {},
-      };
-
-      sessionPool.getSession = function() {
-        return Promise.resolve(SESSION);
-      };
 
-      sessionPool.release = common.util.noop;
-    });
+      it('should create a txn object', function(done) {
+        const fakeOptions = {};
 
-    it('should get a session', function(done) {
-      sessionPool.getSession = function() {
-        setImmediate(done);
-        return Promise.resolve(SESSION);
-      };
+        fakeSession.transaction = function(options) {
+          assert.strictEqual(options, fakeOptions);
+          setImmediate(done);
+          return fakeTxn;
+        };
 
-      sessionPool.request(CONFIG, assert.ifError);
-    });
+        sessionPool.createTransaction_(fakeSession, fakeOptions);
```

Comment:
We're also missing a test that checks to see if the transaction was started correctly via `begin()`

Reply:
Fixed it